### PR TITLE
Refactor syncer state and make notes for future parallelization

### PR DESCRIPTION
**Pull request by danielhlarkin**
This doesn't do any of the heavy lifting yet for the parallelization of the initial syncer, but rather
1) makes some notes about which methods in particular need to be made thread/worker-safe (or verified that they are already so) and
2) refactors a lot of the private or protected state from the various base classes into structs that both handle some of the data manipulation and can be passed more easily to functions in preparation for parallel function.

Basically it looks like a whole bunch of pedantic refactoring since it stops short of actually doing the thing. Some of it might not be 100% necessary, and there's probably still more to be done, but I think these were some of the important first steps and my first guess as to what will need to be done. We can definitely clean it up a bit later when we continue work here.

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/167/

**Comment by jsteemann** [comment]
Windows compilation failed:
```
In file included from /work/ArangoDB/arangod/Replication/utilities.cpp:25:0:
/work/ArangoDB/arangod/Replication/utilities.h:169:22: error: 'Builder' is not a member of 'arangodb::velocypack'
 Result parseResponse(velocypack::Builder&, httpclient::SimpleHttpResult const*);
                      ^~~~~~~~~~
/work/ArangoDB/arangod/Replication/utilities.h:169:42: error: expected primary-expression before ',' token
 Result parseResponse(velocypack::Builder&, httpclient::SimpleHttpResult const*);
                                          ^
/work/ArangoDB/arangod/Replication/utilities.h:169:73: error: expected primary-expression before 'const'
 Result parseResponse(velocypack::Builder&, httpclient::SimpleHttpResult const*);
                                                                         ^~~~~
/work/ArangoDB/arangod/Replication/utilities.cpp: In member function 'arangodb::Result arangodb::replutils::BarrierInfo::create(arangodb::replutils::Connection&, TRI_voc_tick_t)':
/work/ArangoDB/arangod/Replication/utilities.cpp:256:51: error: no match for call to '(arangodb::Result) ({anonymous}::VPackBuilder&, std::unique_ptr<arangodb::httpclient::SimpleHttpResult>::pointer)'
   Result r = parseResponse(builder, response.get());
                                                   ^
/work/ArangoDB/arangod/Replication/utilities.cpp: In member function 'arangodb::Result arangodb::replutils::BatchInfo::start(arangodb::replutils::Connection&, arangodb::replutils::ProgressInfo&)':
/work/ArangoDB/arangod/Replication/utilities.cpp:337:51: error: no match for call to '(arangodb::Result) ({anonymous}::VPackBuilder&, std::unique_ptr<arangodb::httpclient::SimpleHttpResult>::pointer)'
   Result r = parseResponse(builder, response.get());
                                                   ^
/work/ArangoDB/arangod/Replication/utilities.cpp: In constructor 'arangodb::replutils::MasterInfo::MasterInfo(const arangodb::ReplicationApplierConfiguration&)':
/work/ArangoDB/arangod/Replication/utilities.cpp:433:34: error: 'const class arangodb::ReplicationApplierConfiguration' has no member named '_force32mode'
     : _force32mode{applierConfig._force32mode} {}
                                  ^~~~~~~~~~~~
/work/ArangoDB/arangod/Replication/utilities.cpp:433:46: error: cannot convert '<brace-enclosed initializer list>' to 'bool' in initialization
     : _force32mode{applierConfig._force32mode} {}
                                              ^
/work/ArangoDB/arangod/Replication/utilities.cpp: In member function 'arangodb::Result arangodb::replutils::MasterInfo::getState(arangodb::replutils::Connection&, bool)':
/work/ArangoDB/arangod/Replication/utilities.cpp:468:51: error: no match for call to '(arangodb::Result) ({anonymous}::VPackBuilder&, std::unique_ptr<arangodb::httpclient::SimpleHttpResult>::pointer)'
   Result r = parseResponse(builder, response.get());
                                                   ^
/work/ArangoDB/arangod/Replication/utilities.cpp: In function 'arangodb::Result arangodb::replutils::parseResponse(arangodb::velocypack::Builder&, const arangodb::httpclient::SimpleHttpResult*)':
/work/ArangoDB/arangod/Replication/utilities.cpp:528:66: error: 'arangodb::Result arangodb::replutils::parseResponse(arangodb::velocypack::Builder&, const arangodb::httpclient::SimpleHttpResult*)' redeclared as different kind of symbol
                      httpclient::SimpleHttpResult const* response) {
                                                                  ^
In file included from /work/ArangoDB/arangod/Replication/utilities.cpp:25:0:
/work/ArangoDB/arangod/Replication/utilities.h:169:8: note: previous declaration 'arangodb::Result arangodb::replutils::parseResponse'
 Result parseResponse(velocypack::Builder&, httpclient::SimpleHttpResult const*);
        ^~~~~~~~~~~~~
```

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/170/
